### ISSUE TITLE
Inconsistent error reporting with noImplicitAny on enum + namespace merges

### ISSUE THREAD

**Issue by paul-go** [Microsoft/TypeScript#29111]
**TypeScript Version:**  3.3.0-dev.201xxxxx

**Search Terms:** noImplicitAny, iterators, enum, namespace, merge

**Code**

```ts
enum Foo1
{
	a = "aa"
}

namespace Foo1
{
	export function *generator() // <-- Reports an error ... ok
	{
		for (const value of Object.values(Foo1))
			yield value;
	}
	
	export function caller(value: string) // <-- Reports an error ... ok
	{
		for (const value of generator())
			return true;
	
		return false;
	}
}


enum Foo2
{
	a = "aa"
}

namespace Foo2
{
	// No errors reported in this function, even though 
	// the function body is identical to the above.
	export function *generator()
	{
		for (const value of Object.values(Foo2))
			yield value;
	}
	
	export function caller(value: string)
	{
		
	}
}
```

**Expected behavior:** Consistent error reporting 

**Actual behavior:** Two otherwise equivalent code blocks are showing a discrepancy in the reporting of errors due to the presence of (what I believe is) a non-affecting code block.

**Playground Link:** 
https://www.typescriptlang.org/play/#src=enum%20Foo1%0D%0A%7B%0D%0A%09a%20%3D%20%22aa%22%0D%0A%7D%0D%0A%0D%0Anamespace%20Foo1%0D%0A%7B%0D%0A%09export%20function%20*generator()%20%2F%2F%20%3C--%20Reports%20an%20error%20...%20ok%0D%0A%09%7B%0D%0A%09%09for%20(const%20value%20of%20Object.values(Foo1))%0D%0A%09%09%09yield%20value%3B%0D%0A%09%7D%0D%0A%09%0D%0A%09export%20function%20caller(value%3A%20string)%20%2F%2F%20%3C--%20Reports%20an%20error%20...%20ok%0D%0A%09%7B%0D%0A%09%09for%20(const%20value%20of%20generator())%0D%0A%09%09%09return%20true%3B%0D%0A%09%09%0D%0A%09%09return%20false%3B%0D%0A%09%7D%0D%0A%7D%0D%0A%0D%0Aenum%20Foo2%0D%0A%7B%0D%0A%09a%20%3D%20%22aa%22%0D%0A%7D%0D%0A%0D%0Anamespace%20Foo2%0D%0A%7B%0D%0A%09%2F%2F%20No%20errors%20reported%20in%20this%20function%2C%20even%20though%20%0D%0A%09%2F%2F%20the%20function%20body%20is%20identical%20to%20the%20above.%0D%0A%09export%20function%20*generator()%0D%0A%09%7B%0D%0A%09%09for%20(const%20value%20of%20Object.values(Foo2))%0D%0A%09%09%09yield%20value%3B%0D%0A%09%7D%0D%0A%09%0D%0A%09export%20function%20caller(value%3A%20string)%0D%0A%09%7B%0D%0A%09%09%0D%0A%09%7D%0D%0A%7D%0D%0A%0D%0A%2F%2F%20Dropped%20in%20from%20lib.d.ts%2C%20can%20remove%20if%20there%0D%0A%2F%2F%20is%20some%20way%20to%20enable%20lib.es2017.object.d.ts%0D%0A%2F%2F%20in%20the%20playgroun%0D%0Ainterface%20ObjectConstructor%20%7B%0D%0A%20%20%20%20values%3CT%3E(o%3A%20%7B%20%5Bs%3A%20string%5D%3A%20T%20%7D%20%7C%20%20ArrayLike%3CT%3E)%3A%20T%5B%5D%3B%0D%0A%20%20%20%20values(o%3A%20%7B%7D)%3A%20any%5B%5D%3B%0D%0A%7D

**Related Issues:**

**Comment by weswigham** [comment]
In `Foo1` the type of `generator` depends on its yield which depends on the result of `Object.values(Foo1)`, which depends on the types of the values of the properties of `Foo1`, which depends on the signature of `caller`, which in turn depends on its internal call to `generator()` (since we have to check that it's not a zero-length array which would make the block after it unreachable and the function only return `false`!) which is a cyclic dependency. So we issue an error and you need to break the dependency by adding a type annotation somewhere (anywhere within the cycle will do).

In `Foo2`, `caller` doesn't contain a `generator` call, so there's no cyclic type dependency and no issues inferring any types.

**Comment by paul-go** [comment]
Well there's a nice little rabbit hole. It really did look like a bug rather than expected behavior from the standpoint of a person who isn't studying TS language semantics all day :-)
